server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/projection/cassandra/CassandraMailQueueProjection.java

```diff
@@ -0,0 +1,60 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.queue.rabbitmq.projection.cassandra;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+
+import org.apache.james.queue.rabbitmq.projection.api.MailQueueProjection;
+import org.apache.james.queue.rabbitmq.projection.api.model.MailQueueItem;
+import org.apache.james.queue.rabbitmq.projection.cassandra.model.CassandraMailQueueItem;
+
+public class CassandraMailQueueProjection implements MailQueueProjection {
+
+    private final FirstEnqueuedMailDao firstEnqueuedMailDao;
+    private final EnqueuedMailsDao enqueuedMailsDao;
+
+    @Inject
+    public CassandraMailQueueProjection(FirstEnqueuedMailDao firstEnqueuedMailDao, EnqueuedMailsDao enqueuedMailsDao) {
+        this.firstEnqueuedMailDao = firstEnqueuedMailDao;
+        this.enqueuedMailsDao = enqueuedMailsDao;
+    }
+
+    @Override
+    public Stream<? extends MailQueueItem> browse(String queueName) {
+        return firstEnqueuedMailDao
+            .findFirstEnqueuedItem(queueName)
+            .thenCompose(maybeMailQueueItem -> maybeMailQueueItem
+                .map(firstEnqueued -> findEnqueuedMailFrom(queueName, firstEnqueued.getTimeRangeStart()))
+                .orElse(emptyStreamFuture()))
+            .join();
+    }
+
+    private CompletableFuture<Stream<CassandraMailQueueItem>> emptyStreamFuture() {
+        return CompletableFuture.completedFuture(Stream.empty());
+    }
+
+    private CompletableFuture<Stream<CassandraMailQueueItem>> findEnqueuedMailFrom(String queueName, Long timeRangeStartFrom) {
```

Comment:
You should : 

 - 1. Read FirstEnqueuedMailDao to find a timestamp that minorate all enqueued email
 - 2. Round it down to the first slide. (eg if a slide size is one hour, and my first enqueued mail is 3pm57'45 I need to read from 'slidee 3pm'
 - 3. Read all content from this slide, and only keep non dequeued emails.

Please note that we can opportunistically update the first enqueued timestamp upon browse (value is known)

Reply:
the calculating slide stuff is messy, I keep it to the next implementation session.